### ISSUE TITLE
Delayed Slack Slash Commands

### ISSUE THREAD

**Issue by mattevolve** [botman/botman#720]
- BotMan Version: 2.2
- PHP Version: 7.2
- Messaging Service(s): Slack
- Slack Driver: 2.1

### Description:
What do I have to do to get a delayed response into my Slack workspace? I'm submitting a Slack slash command but there is a delay between the logic and the response and I cannot guarantee the response will be ready and returned within Slack's allotted 3000ms. I have a laravel 5.5 installation with botman and all of my logic is currently within the `routes/botman.php` file.

```
$botman->hears('/command {var}', function ($bot, $var) {
    $attachments = slackCommand('command', $var); // custom logic happens here
    $bot->sendPayload([
        'attachments' => buildPayload($attachments) // attachment data compiled via custom function
    ]);
});
```

Apologies if this is already documented but I can't figure this out... What must I alter to send a delayed response? Any logic within the closure that delays the response past 3000ms causes this Slack response:

> Darn - that slash command didn't work (error message: `Timeout was reached`). Manage the command at BotName.

Is there a way botman could automatically handle delayed responses when detecting a slash command using the slack driver `isSlashCommand` function?

### TL;DR:
If I have understood the botman/Slack documentation correctly, when botman listens for a slack command the response should contain a `response_url=https://hooks.slack.com/commands/1234/5678` key and value. I need to take the `response_url` and post to it after my application has performed its logic but this needs to happen separately from the initial `$botman->hears()` closure? Am I thinking about this in the right way and do I need to dispatch a laravel job or something?

https://api.slack.com/slash-commands#responding_to_a_command

### Update
@mpociot I noticed @ajmariduena was experiencing this issue here: https://github.com/botman/botman/issues/667

I have updated my code to dispatch a job but the job handle doesn't seem able to send the bot reply. I need to send a response to the `response_url` webhook but I don't believe this is accessible from the `IncomingMessage` object. Does botman handle this out of the box? I tried the suggestions by @ajmariduena but this doesn't appear to resolve the issue: https://github.com/botman/botman/issues/667#issuecomment-356339681.

#### routes/botman.php
```
$botman->hears('/command {var}', function ($bot, $var) {
    \App\Jobs\SlowSlackResponse::dispatch($bot);
    $bot->reply('Give me a second while I fetch information for '.$var.'.');
});
```

#### App/Jobs/SlowSlackResponse.php
```
<?php

namespace App\Jobs;

use Illuminate\Bus\Queueable;
use Illuminate\Queue\SerializesModels;
use Illuminate\Queue\InteractsWithQueue;
use Illuminate\Contracts\Queue\ShouldQueue;
use Illuminate\Foundation\Bus\Dispatchable;
use BotMan\BotMan\BotMan;

class SlowSlackResponse implements ShouldQueue
{
    use Dispatchable, InteractsWithQueue, Queueable, SerializesModels;

    protected $bot;

    /**
     * Create a new job instance.
     *
     * @return void
     */
    public function __construct(BotMan $bot)
    {
        $this->bot = $bot;
    }

    /**
     * Execute the job.
     *
     * @return void
     */
    public function handle()
    {
        $this->bot->reply('Here is the delayed response.');
    }
}
```

#### Job Response
```
[2018-02-21 16:10:26][738] Processing: App\Jobs\SlowSlackResponse
{"text":"Here is the delayed response."}[2018-02-21 16:10:26][738] Processed:  App\Jobs\SlowSlackResponse
```
Queue driver: database

**Comment by mpociot** [comment]
Hm yeah, in this case, the problem is that you're using the webhook integration which does not POST a reponse to a Slack API, but instead is a simple JSON response.
That's why you see the JSON when the job gets processed.

Probably the only way around this would be: Use the Slack event API or use the SlackRTM integration using `php artisan botman:listen-on-slack`.
In this case you wouldn't even need the job, as the bot runs on a socket in the background.
